### DISCUSSION TITLE
How to setup multi-view port support

### DISCUSSION THREAD

**Discussion by immutheprogrammer** [ocornut/imgui#4112] [Old/Archived discussions]
I can't figure out how to set up multi-viewport support does anyone have any tips?

**Comment by PathogenDavid** [comment]
I'm assuming you're asking in relation to [this project](https://github.com/immutheprogrammer/Run-Game-Engine).

It's not immediately apparent to me whether you're using the provided backends or a custom one. It seems like you're using a built-in one but you started writing a custom one?

# Built-in backends

If you're using the built-in backends, I think you should just need to enable it in [`ImGuiLayer.initImGui`](https://github.com/immutheprogrammer/Run-Game-Engine/blob/ffa500888299b152be4f47cef0c4207789583c51/src/src/main/java/run/im_gui/ImGuiLayer.java#L48) by adding:

```java
io.addConfigFlags(ImGuiConfigFlags.ViewportsEnable);    // Enable Multi-Viewport / Platform Windows
```

you might also want to add

```java
io.addConfigFlags(ImGuiConfigFlags.DockingEnable);      // Enable Docking
```

to enable docking.

# Custom backends

If you're making your own backend, I highly recommend writing a backend without multi-viewport support first.

From there, to add multiple viewport support the main thing you need to do is fill out the callbacks in [`ImGuiPlatformIO`](https://github.com/ocornut/imgui/blob/31290807dc169d4ee62b16f42bcb1d9a294be93b/imgui.h#L2937-L2998). I've also found it useful to study the differences between the backends in the main and docking branches.

I cannot give you specific advice on how to do this with imgui-java or whether it's even possible given their bindings. I'd suggest [asking them directly](https://github.com/SpaiR/imgui-java) if you run into problems.

I don't believe there's really any documentation yet for implementing a custom platform backend because the feature is still in development.
